LanguageParser.hs

```diff
@@ -1,40 +1,414 @@
+            enterP' = Enter <$> (P.char '[' *> P.space *> P.string' "Enter" *>
+                      P.space1 *> (P.try double <|> single) <* P.space <*
+                      P.char ']' <* P.space)
+                      P.<?> "an enter scene"
+
+exitP :: Parser (Statement, Annotation)
+exitP = swap <$> liftA2 (,) enterExitAnnotationP exitP'
+         where
+           exitP' = Exit <$> (P.char '[' *> P.space *> P.string' "Exit" *>
+                    P.space1 *> oneOfCharacterNames <* P.space <*
+                    P.char ']' <* P.space)
+                    P.<?> "an exit scene"
+
+exeuntP :: Parser (Statement, Annotation)
+exeuntP = swap <$> liftA2 (,) enterExitAnnotationP exeuntP'
+           where
+             exeuntP' = Exeunt <$> (P.char '[' *> P.space *> P.string'
+                        "Exeunt" *> (P.try (P.space1 *> double) <|> none) <*
+                        P.space <* P.char ']' <* P.space)
+                        P.<?> "an exeunt scene"
+
+none :: Parser [String]
+none = const [] <$> P.takeP Nothing 0
+
+single :: Parser [String]
+single = (:[]) <$> oneOfCharacterNames <* P.space
+
+-- TODO: Handle three, four, five, etc [though not necessary].
+double :: Parser [String]
+double = liftA2 (\a b -> [a, b])
+         (oneOfCharacterNames <* P.space1 <* P.string "and" <* P.space1)
+         (oneOfCharacterNames <* P.space)
+
+lineP :: Parser [(Statement, Annotation)]
+lineP = liftA2 combine (Line <$> oneOfCharacterNames <* P.char ':' <* P.space)
+        listOfSentenceP <* P.space
+        P.<?> "a line"
+        where
+          combine line senAnnList = (\(s, a) -> (line s, a)) <$> senAnnList
+
+listOfSentenceP :: Parser [(Sentence, Annotation)]
+listOfSentenceP = some sentenceP
+
+ifSoP :: Parser Sentence
+ifSoP = IfSo <$> (P.string' "If" *> P.space1 *> P.string' "so" *>
+                 P.space *> P.char ',' *> P.space *> sentenceP')
+
+ifNotP :: Parser Sentence
+ifNotP = IfNot <$> (P.string' "If" *> P.space1 *> P.string' "not" *>
+                   P.space *> P.char ',' *> P.space *> sentenceP')
+
+outputNumberP :: Parser Sentence
+outputNumberP = constP OutputNumber
+                (P.string' "Open" <* P.space1 <* oneOfSecondPersonPos <*
+                P.space1 <* P.string' "heart")
+
+outputCharacterP :: Parser Sentence
+outputCharacterP = constP OutputCharacter
+                   (P.string' "Speak" <* P.space1 <* oneOfSecondPersonPos <*
+                   P.space1 <* P.string' "mind")
+
+inputNumberP :: Parser Sentence
+inputNumberP = constP InputNumber
+               (P.string' "Listen" <* P.space1 <* P.string "to" <* P.space1 <*
+               oneOfSecondPersonPos <* P.space1 <* P.string' "heart")
+
+inputCharacterP :: Parser Sentence
+inputCharacterP = constP InputCharacter
+                  (P.string' "Open" <* P.space1 <* oneOfSecondPersonPos <*
+                  P.space1 <* P.string' "mind")
+
+declarationP :: Parser Sentence
+declarationP =
+  P.try decVarient1 <|> P.try decVarient2 <|> decVarient3
```

Comment:
less repeated code if you do `Declaration <$> (P.try decVarient1 <|> P.try decVarient2 <|> decVarient3) *> expressionP` and then change `decVarient`s appropriately.